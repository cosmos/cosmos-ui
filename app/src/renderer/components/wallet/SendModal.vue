<template>
  <action-modal
    ref="actionModal"
    :submit-fn="submitForm"
    :validate="validateForm"
    title="Send"
    submission-error-prefix="Sending tokens failed"
  >
    <tm-form-group
      :error="$v.denom.$dirty && $v.denom.$invalid"
      field-id="send-denomination"
      field-label="Denomination"
    >
      <tm-field
        id="send-denomination"
        v-model.number="$v.denom.$model"
        type="text"
        readonly
      />
      <tm-form-msg
        v-if="$v.denom.$error && !$v.denom.required"
        name="Denomination"
        type="required"
      />
    </tm-form-group>

    <tm-form-group
      :error="$v.address.$error && $v.address.$invalid"
      field-id="send-address"
      field-label="Send To"
    >
      <tm-field
        v-focus
        id="send-address"
        v-model.number="$v.address.$model"
        type="text"
        placeholder="Address"
      />
      <tm-form-msg
        v-if="$v.address.$error && !$v.address.required"
        name="Address"
        type="required"
      />
      <tm-form-msg
        v-else-if="$v.address.$error && !$v.address.bech32Validate"
        :body="bech32error"
        name="Address"
        type="bech32"
      />
    </tm-form-group>

    <tm-form-group
      :error="$v.amount.$error && $v.amount.$invalid"
      field-id="send-amount"
      field-label="Amount"
    >
      <tm-field
        id="send-amount"
        :max="max"
        :min="max ? 1 : 0"
        v-model.number="$v.amount.$model"
        type="number"
        placeholder="Amount"
      />
      <tm-form-msg
        v-if="$v.amount.$error && !$v.amount.required"
        name="Amount"
        type="required"
      />
      <tm-form-msg
        v-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <tm-form-msg
        v-else-if="$v.amount.$error && !$v.amount.integer"
        name="Amount"
        type="integer"
      />
    </tm-form-group>
  </action-modal>
</template>

<script>
import b32 from "scripts/b32"
import { required, between, integer } from "vuelidate/lib/validators"
import { mapActions, mapGetters } from "vuex"
import TmBtn from "common/TmBtn"
import TmFieldGroup from "common/TmFieldGroup"
import TmFormGroup from "common/TmFormGroup"
import TmFormStruct from "common/TmFormStruct"
import TmPage from "common/TmPage"
import TmPart from "common/TmPart"
import TmField from "common/TmField"
import TmFormMsg from "common/TmFormMsg"
import TmBalance from "common/TmBalance"
import FieldAddon from "common/TmFieldAddon"
import ToolBar from "common/ToolBar"
import ActionModal from "common/ActionModal"

export default {
  name: `send-modal`,
  components: {
    TmBalance,
    TmBtn,
    TmField,
    FieldAddon,
    TmFieldGroup,
    TmFormGroup,
    TmFormMsg,
    TmFormStruct,
    TmPage,
    TmPart,
    ToolBar,
    ActionModal
  },
  data: () => ({
    bech32error: null,
    address: ``,
    amount: ``,
    denom: ``
  }),
  computed: {
    ...mapGetters([`wallet`]),
    max() {
      let denom = this.wallet.balances.find(b => b.denom === this.denom)
      return (denom && denom.amount) || 0
    },
    denominations() {
      return this.wallet.balances.map(i => ({
        key: i.denom.toUpperCase(),
        value: i.denom
      }))
    }
  },
  mounted() {
    if (this.denom) {
      this.denom = this.denom
    }
  },
  methods: {
    ...mapActions([`sendTx`]),
    open(denom) {
      this.denom = denom
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    async submitForm(submitType, password) {
      const amount = +this.amount
      const address = this.address
      const denom = this.denom
      const type = `send`

      await this.sendTx({
        type,
        submitType,
        password: password,
        to: address,
        amount: [{ denom, amount: amount.toString() }]
      })

      this.$store.commit(`notify`, {
        title: `Successfully Sent`,
        body: `Successfully sent ${amount} ${denom} to ${address}`
      })
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        this.bech32error = null
        return true
      } catch (error) {
        this.bech32error = error.message
        return false
      }
    }
  },
  validations() {
    return {
      address: {
        required,
        bech32Validate: this.bech32Validate
      },
      amount: {
        required,
        integer,
        between: between(this.max ? 1 : 0, this.max)
      },
      denom: { required }
    }
  }
}
</script>
