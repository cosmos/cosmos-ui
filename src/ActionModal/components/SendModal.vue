<template>
  <ActionModal
    id="send-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="amount"
    title="Send"
    submission-error-prefix="Sending tokens failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    @close="clear"
  >
    <TmFormGroup
      :error="$v.denom.$dirty && $v.denom.$invalid"
      class="action-modal-form-group"
      field-id="send-denomination"
      field-label="Denomination"
    >
      <TmField
        id="send-denomination"
        :value="viewDenom($v.denom.$model)"
        type="text"
        readonly
      />
      <TmFormMsg
        v-if="$v.denom.$error && !$v.denom.required"
        name="Denomination"
        type="required"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.address.$error && $v.address.$invalid"
      class="action-modal-form-group"
      field-id="send-address"
      field-label="Send To"
    >
      <TmField
        id="send-address"
        v-model.number="$v.address.$model"
        v-focus
        type="text"
        placeholder="Address"
        @keyup.enter.native="refocusOnAmount"
      />
      <TmFormMsg
        v-if="$v.address.$error && !$v.address.required"
        name="Address"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.address.$error && !$v.address.bech32Validate"
        name="Address"
        type="bech32"
      />
    </TmFormGroup>
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix max-button">{{ viewDenom(denom) }}</span>
      <TmFieldGroup>
        <TmField
          id="amount"
          ref="amount"
          v-model="amount"
          class="tm-field-addon"
          placeholder="Amount"
          type="number"
          @keyup.enter.native="enterPressed"
        />
        <TmBtn
          type="button"
          class="secondary addon-max"
          value="Set Max"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <TmFormMsg
        v-if="currentBalance === 0"
        :msg="`doesn't have any ${viewDenom(denom)}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
      <TmFormMsg
        v-else-if="isMaxAmount()"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc max-message"
      />
    </TmFormGroup>
    <TmBtn
      v-if="editMemo === false"
      id="edit-memo-btn"
      value="Edit Memo"
      type="secondary"
      @click.native="showMemo()"
    />
    <TmFormGroup
      v-else
      id="memo"
      :error="$v.memo.$error && $v.memo.$invalid"
      class="action-modal-group"
      field-id="memo"
      field-label="Memo"
    >
      <TmField
        id="memo"
        v-model="memo"
        type="text"
        placeholder="Let everyone know how much you love Lunie"
        @keyup.enter.native="enterPressed"
      />
      <TmFormMsg
        v-if="$v.memo.$error && !$v.memo.maxLength"
        name="Memo"
        type="maxLength"
        :max="max_memo_characters"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import gql from "graphql-tag"
import b32 from "scripts/b32"
import { required, between, decimal, maxLength } from "vuelidate/lib/validators"
import { uatoms, atoms, viewDenom, SMALLEST } from "src/scripts/num"
import { mapGetters } from "vuex"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import transaction from "../utils/transactionTypes"

function reverseDenomLookup(denom) {
  if (denom === 'ATOM') {
    return 'uatom'
  }
  return denom.toLowerCase()
}

const defaultMemo = "(Sent via Lunie)"

export default {
  name: `send-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmFormGroup,
    TmFormMsg,
    ActionModal,
    TmBtn
  },
  data: () => ({
    address: ``,
    amount: null,
    denom: ``,
    memo: defaultMemo,
    max_memo_characters: 256,
    editMemo: false,
    balance: []
  }),
  computed: {
    ...mapGetters([`network`]),
    ...mapGetters({ userAddress: `address` }),
    currentBalance() {
      const denom = this.balance.find(b => b.denom === this.denom)
      return (denom && denom.amount) || 0
    },
    transactionData() {
      return {
        type: transaction.SEND,
        toAddress: this.address,
        amounts: [
          {
            amount: uatoms(+this.amount),
            denom: reverseDenomLookup(this.denom)
          }
        ],
        memo: this.memo
      }
    },
    notifyMessage() {
      return {
        title: `Successful Send`,
        body: `Successfully sent ${+this.amount} ${viewDenom(this.denom)}s to ${
          this.address
        }`
      }
    }
  },
  methods: {
    viewDenom,
    open(denom) {
      this.denom = denom
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.address = undefined
      this.amount = undefined
      this.editMemo = false
      this.memo = defaultMemo
      this.sending = false
    },
    setMaxAmount() {
      this.amount = this.currentBalance
    },
    isMaxAmount() {
      if (this.currentBalance === 0) {
        return false
      } else {
        return parseFloat(this.amount) === parseFloat(this.currentBalance)
      }
    },
    bech32Validate(param) {
      try {
        b32.decode(param)
        return true
      } catch (error) {
        return false
      }
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    refocusOnAmount() {
      this.$refs.amount.$el.focus()
    },
    showMemo() {
      this.memo = ``
      this.editMemo = true
    }
  },
  validations() {
    return {
      address: {
        required,
        bech32Validate: this.bech32Validate
      },
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, this.currentBalance)
      },
      denom: { required },
      memo: {
        maxLength: maxLength(this.max_memo_characters)
      }
    }
  },
  apollo: {
    balance: {
      query: gql`
        query balance($networkId: String!, $address: String!) {
          balance(networkId: $networkId, address: $address) {
            amount
            denom
          }
        }
      `,
      skip() {
        return !this.userAddress
      },
      variables() {
        return {
          networkId: this.network,
          address: this.userAddress
        }
      },
      update: data => {
        return data.balance
      }
    }
  }
}
</script>
<style scoped>
#edit-memo-btn {
  margin-top: 2.4rem;
}
</style>
