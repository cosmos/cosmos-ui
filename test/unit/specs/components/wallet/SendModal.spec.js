import { shallowMount, createLocalVue } from "@vue/test-utils"
import Vuelidate from "vuelidate"
import SendModal from "renderer/components/wallet/SendModal"

describe(`SendModal`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive(`focus`, () => { })

  let wrapper, $store

  const balances = [
    {
      denom: `STAKE`,
      amount: 10000000000
    },
    {
      denom: `fermion`,
      amount: 2300
    }
  ]
  const getters = {
    wallet: {
      loading: false,
      denoms: [`fermion`, `gregcoin`, `mycoin`, `STAKE`],
      balances
    },
    connected: true,
    session: { signedIn: true }
  }

  beforeEach(async () => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters
    }

    wrapper = shallowMount(SendModal, {
      localVue,
      propsData: {
        denom: `STAKE`
      },
      mocks: {
        $store
      },
      sync: false
    })

    wrapper.vm.$refs.actionModal = {
      submit: cb => cb(),
      open: jest.fn()
    }
  })

  it(`should display send modal form`, async () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      address: `cosmos1address`,
      amount: 10
    }
    SendModal.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.address).toBe(``)
    expect(self.amount).toBe(0)
  })

  describe(`validation`, () => {
    it(`should show address required error`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: ``,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
    it(`should show bech32 error when address length is too short`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: `asdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })

    it(`should show bech32 error when address length is too long`, async () => {
      wrapper.setData({
        denom: `STAKE`,
        address: `asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf`,
        amount: 2
      })
      wrapper.vm.validateForm()
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$v.$error).toBe(true)
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
    it(`should show bech32 error when alphanumeric is wrong`, async () => {
      wrapper.setData({
        address: ``
      })
      expect(wrapper.vm.validateForm()).toBe(false)
      await wrapper.vm.$nextTick()
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })

  describe(`simulateForm`, () => {
    it(`should simulate transaction to estimate gas used`, async () => {
      const estimate = 1234567
      const $store = { dispatch: jest.fn(() => estimate) }
      const res = await SendModal.methods.simulateForm.call(
        {
          $store,
          amount: 10,
          address: `cosmos1address`,
          denom: `uatom`,
        }
      )

      expect($store.dispatch).toHaveBeenCalledWith(`simulateTx`,
        {
          type: `send`,
          to: `cosmos1address`,
          amount: [{ amount: `10000000`, denom: `uatom` }]
        }
      )
      expect(res).toBe(estimate)
    })
  })

  describe(`submitForm`, () => {
    it(`submits a transfer transaction`, async () => {
      const $store = { commit: jest.fn() }
      const sendTx = jest.fn()
      const gas = `1234567`
      const gasPrice = 2.5e-8
      const gas_prices = [{ denom: `uatom`, amount: `0.025` }]

      await SendModal.methods.submitForm.call(
        {
          amount: 10,
          address: `cosmos1address`,
          denom: `uatom`,
          bondDenom: `uatom`,
          $store,
          sendTx
        },
        gas, gasPrice, ``, `ledger`
      )

      expect(sendTx).toHaveBeenCalledWith(
        {
          type: `send`,
          to: `cosmos1address`,
          amount: [{ amount: `10000000`, denom: `uatom` }],
          gas,
          gas_prices,
          submitType: `ledger`,
          password: ``
        }
      )

      expect($store.commit).toHaveBeenCalledWith(`notify`,
        {
          body: `Successfully sent 10 ATOMs to cosmos1address`,
          title: `Successful Send`
        }
      )
    })
  })
  it(`validates bech32 addresses`, () => {
    expect(
      wrapper.vm.bech32Validate(`cosmos1x7wzdumfj8pncd99mqc0mkqfrrps3l3pjz8tk6`)
    ).toBe(true)
    expect(
      wrapper.vm.bech32Validate(`cosmos15ky9du8a2wlstz6fpx3p4mqpjyrm5ctpesxxn9`)
    ).toBe(false)
  })
})
