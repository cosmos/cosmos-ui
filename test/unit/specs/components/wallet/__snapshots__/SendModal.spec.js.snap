// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendModal disables sending if not connected 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="2300"
          min="1"
          placeholder="Amount"
          type="number"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Connecting...
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`SendModal has the expected html structure 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="2300"
          min="1"
          placeholder="Amount"
          type="number"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`SendModal should show address required error 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <div
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is required
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="0"
          min="0"
          placeholder="Amount"
          type="number"
        />
         
        <!---->
         
        <div
          class="tm-form-msg sm tm-form-msg--error"
        >
          Amount must be between 0 and 0
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`SendModal should show bech32 error when address length is too long 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <div
          body="No separator character for asdfasdfasdfasdfasdfasdfasdfasdfasdfasdfasdf"
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is invalid bech32
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="2300"
          min="1"
          placeholder="Amount"
          type="number"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`SendModal should show bech32 error when address length is too short 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <div
          body="asdf too short"
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is invalid bech32
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="2300"
          min="1"
          placeholder="Amount"
          type="number"
        />
         
        <!---->
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`SendModal should show bech32 error when alphanumeric is wrong 1`] = `
<div
  class="action-modal"
  name="slide-fade"
>
  <div
    class="action-modal-header"
  >
    <img
      class="icon action-modal-atom"
      src="~assets/images/cosmos-logo.png"
    />
    <span
      class="action-modal-title"
    >
      Send
    </span>
     
    <div
      class="action-modal-icon action-modal-close"
      id="closeBtn"
    >
      <i
        class="material-icons"
      >
        close
      </i>
    </div>
  </div>
   
  <div
    class="action-modal-form"
  >
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-denomination"
      >
        Denomination
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-denomination"
          readonly="readonly"
          type="text"
        />
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-address"
      >
        Send To
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-address"
          placeholder="Address"
          type="text"
        />
         
        <div
          class="tm-form-msg sm tm-form-msg--error"
        >
          Address is required
        </div>
      </div>
    </div>
     
    <div
      class="tm-form-group tm-form-group--error"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="send-amount"
      >
        Amount
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="send-amount"
          max="2300"
          min="1"
          placeholder="Amount"
          type="number"
        />
         
        <div
          class="tm-form-msg sm tm-form-msg--error"
        >
          Amount is required
        </div>
         
        <!---->
      </div>
    </div>
     
    <div
      class="tm-form-group"
    >
      <!---->
      <label
        class="tm-form-group__label"
        for="password"
      >
        Password
      </label>
       
      <div
        class="tm-form-group__field"
      >
        <input
          class="tm-field"
          id="password"
          placeholder="Password"
          type="password"
        />
         
        <!---->
      </div>
    </div>
     
  </div>
   
  <div
    class="action-modal-footer"
  >
    <div>
      <button
        class="tm-btn"
        disabled="disabled"
        id="send-btn"
      >
        <span
          class="tm-btn__container tm-btn--primary"
        >
          <!---->
          <span
            class="tm-btn__value"
          >
            Send Tokens
          </span>
        </span>
      </button>
    </div>
     
    <!---->
  </div>
</div>
`;
